### PR TITLE
feat: optional cascade delete resources when deleting workspace

### LINKED ISSUE #3192: Choose whether to cascade delete related resources when deleting workspace
**What's it about?**

Choose whether to cascade delete related resources when deleting workspace, avoid accidental deletion of related resources.

/kind feature-request

### PR DESCRIPTION
**Pull request by RolandMa1986**
Signed-off-by: Roland.Ma <rolandma@yunify.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add DeleteOptions flags support for workspace deletion API. So user can choose whether to cascade delete related resources when deleting workspace, avoid accidental deletion of related resources.

This feature depends on Kubefed 0.7.0

**Which issue(s) this PR fixes**:
Fixes #3192

### CODE REVIEW

**Review by LinuxSuRen — APPROVED** [review]
/lgtm

File: pkg/controller/namespace/namespace_controller.go [hunk 1]
```diff
@@ -188,11 +192,18 @@ func (r *Reconciler) bindWorkspace(ctx context.Context, logger logr.Logger, name
 }
 
 func (r *Reconciler) unbindWorkspace(ctx context.Context, logger logr.Logger, namespace *corev1.Namespace) error {
-	if k8sutil.IsControlledBy(namespace.OwnerReferences, tenantv1alpha1.ResourceKindWorkspace, "") {
-		namespace := namespace.DeepCopy()
-		namespace.OwnerReferences = k8sutil.RemoveWorkspaceOwnerReference(namespace.OwnerReferences)
-		logger.V(4).Info("remove owner reference", "workspace", namespace.Labels[constants.WorkspaceLabelKey])
-		if err := r.Update(ctx, namespace); err != nil {
+	if k8sutil.IsControlledBy(namespace.OwnerReferences, tenantv1alpha1.ResourceKindWorkspace, "") || len(namespace.Labels) > 0 {
```
Comment (wansir) [inline pkg/controller/namespace/namespace_controller.go:195]:
```suggestion
	if k8sutil.IsControlledBy(namespace.OwnerReferences, tenantv1alpha1.ResourceKindWorkspace, "") || namespace.Labels[tenantv1alpha1.WorkspaceLabel] != "" {
```

File: pkg/controller/workspacetemplate/workspacetemplate_controller.go [hunk 2]
```diff
@@ -105,16 +107,30 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+				}
+			}
+
 			// remove our finalizer from the list and update it.
 			workspaceTemplate.ObjectMeta.Finalizers = sliceutil.RemoveString(workspaceTemplate.ObjectMeta.Finalizers, func(item string) bool {
 				return item == workspaceTemplateFinalizer
 			})
+			workspaceTemplate.ObjectMeta.Finalizers = sliceutil.RemoveString(workspaceTemplate.ObjectMeta.Finalizers, func(item string) bool {
+				return item == orphanFinalizer
+			})
```
Comment (wansir) [inline pkg/controller/workspacetemplate/workspacetemplate_controller.go:132]:
```suggestion
			// remove our finalizer from the list and update it.
			workspaceTemplate.ObjectMeta.Finalizers = sliceutil.RemoveString(workspaceTemplate.ObjectMeta.Finalizers, func(item string) bool {
				return item == workspaceTemplateFinalizer || item == orphanFinalizer
			})
```

Comment (RolandMa1986) [reply pkg/controller/workspacetemplate/workspacetemplate_controller.go]: Resolved

**Review by LinuxSuRen — APPROVED** [review]
/lgtm

**Comment by yuswift** [comment]
When deleting a federatedworkspace, can we choose if we want to delete the associated namespaces?
@RolandMa1986

**Comment by RolandMa1986** [comment]
> When deleting a federatedworkspace, can we choose if we want to delete the associated namespaces?
> @RolandMa1986

@yuswift, in the Backend API, we just provided the ability of whether or not to delete the associated namespaces.  But can't let the user choose which one can be deleted or not.

**Comment by yuswift** [comment]
> > When deleting a federatedworkspace, can we choose if we want to delete the associated namespaces?
> > @RolandMa1986
> 
> @yuswift, in the Backend API, we just provided the ability of whether or not to delete the associated namespaces. But can't let the user choose which one can be deleted or not.

Sure. It's reasonable. So if I want to delete a federaetedWorkSpace, I can choose if I want to delete the associated namespaces. Is that right? BTW, if I choose not, another user created a federatedWorkSpace, will he see the associated namespaces?

**Comment by RolandMa1986** [comment]
> > > When deleting a federatedworkspace, can we choose if we want to delete the associated namespaces?
> > > @RolandMa1986
> > 
> > 
> > @yuswift, in the Backend API, we just provided the ability of whether or not to delete the associated namespaces. But can't let the user choose which one can be deleted or not.
> 
> Sure. It's reasonable. So if I delete a federaetedWorkSpace, I can choose if I want to delete the associated namespaces. Is that right? BTW, if I choose not, another user created a federatedWorkSpace with the same name, will he see the associated namespaces?
For the second question, the answer is No.  Workspace relations(workspace label) will be removed too. You need re-assign the namespace to another workspace.

**Pull request merged**